teps/0048-pipeline-as-code.md

```diff
@@ -0,0 +1,214 @@
+---
+status: proposed
+title: Pipelines as Code
+creation-date: '2020-01-20'
+last-updated: '2020-01-20'
+authors: [chmouel, afrittoli]
+---
+
+# TEP-0048: Pipeline as Code
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+- [Requirements](#requirements)
+- [Proposal](#proposal)
+  - [Notes/Caveats (optional)](#notescaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [User Experience (optional)](#user-experience-optional)
+  - [Performance (optional)](#performance-optional)
+- [Design Details](#design-details)
+- [Test Plan](#test-plan)
+- [Design Evaluation](#design-evaluation)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (optional)](#infrastructure-needed-optional)
+- [Upgrade &amp; Migration Strategy (optional)](#upgrade--migration-strategy-optional)
+- [References (optional)](#references-optional)
+
+## Summary
+
+The pipelines as code technique emphasizes that the configuration of delivering
+pipelines that build, test and deploy our applications or infrastructure should
+be treated as code; they should be placed under source control and modularized
+in reusable components with automated testing and deployment.
+
+
+## Motivation
+
+The Tekton pipeline definitions are currently dissociated from a source code repository. 
+They may be located in the same repository as they code they should build, test and deploy,
+but Tekton is not aware of them. Tekton expects the pipeline definition to be deployed to a
+kubernetes cluster. Even when the Tekton resources are deployed to the cluster from the
+Repository, it's not easy to test changes to them, as their definition won't be updated until
+they are redeployed to the cluster. 
+
+The reason for the current behaviour is that Tekton is non-prescriptive about how to set up
+CI and CD pipelines. It aims to provide the building blocks for creating such pipelines and it leaves
+it up to the user to decide how to integrate the pipeline in the overall workflow. 
+
+The use case of a single repo, running CI/CD jobs in isolation, is very common and
+we wish to provide dedicated support for it, by providing an easy way for users to
+run pipelines and tasks stored in a specific folder of the repo. The same tool can be used by
+pipeline and task authors for quick testing and experimentation of their code.
+
```

Comment:
something else that might be interesting to mention as motivation is that this is a feature supported by so so many other CI/CD systems

Reply:
@chmouel @FogDong I spent a bit of time looking at some other systems and whether or not most of them allow you to express both pipelines + triggering information as "code" (i.e. use version control as the source of truth) and it seems like most of them do

[In this doc](https://docs.google.com/document/d/1WrT_MTBkkR-bt50qu_fcGE11v9aMzCm262KatxHCZew/edit#) I took a really quick (and maybe inaccurate, plz add any corrections) look at:
* Github actions
* aws codepipeline
* azure pipeline
* jenkins (im the least sure i found the right info for this one)
* circleci
* travis
* harness
* bamboo
* gitlab

I wanted to look at more, including Openshift pipelines ( + @vdemeester @ImJasonH ?) but didnt get a chance!

plz feel free to add to this / expand on it / correct it / use anything useful in this TEP

Reply:
Thanks for doing this @bobcatfish

Yeah a lot of this has been focused on Pull request CI (pre-merge) and not much on CD and post-merge CI.

I definitively think it's in scope, there is Jenkins-X who supports this too with [lighthouse](https://github.com/jenkins-x/lighthouse) which I find interesting. 

We will definitively needs to do a bit of thinking on how to integrate this with triggers.

(and fyi openshift BuildConfig, openshift pipelines doesn't support triggering as code)